napi/src/options.rs

```diff
@@ -0,0 +1,274 @@
+#[derive(Debug, PartialEq, Eq)]
+pub enum EnforceExtension {
+    Auto,
+    Enabled,
+    Disabled,
+}
+
+impl EnforceExtension {
+    pub fn is_auto(&self) -> bool {
+        *self == Self::Auto
+    }
+
+    pub fn is_enabled(&self) -> bool {
+        *self == Self::Enabled
+    }
+
+    pub fn is_disabled(&self) -> bool {
+        *self == Self::Disabled
+    }
+}
+
+/// Alias Value for [ResolveOptions::alias] and [ResolveOptions::fallback].
+/// Use struct because napi don't support structured union now
+#[napi(object)]
+#[derive(Debug, Clone)]
+pub struct Restriction {
+    pub path: Option<String>,
+    pub regex: Option<String>,
+}
+
+/// Tsconfig Options
+///
+/// Derived from [tsconfig-paths-webpack-plugin](https://github.com/dividab/tsconfig-paths-webpack-plugin#options)
+
+#[napi(object)]
+#[derive(Debug, Clone)]
+pub struct TsconfigOptions {
+    /// Allows you to specify where to find the TypeScript configuration file.
+    /// You may provide
+    /// * a relative path to the configuration file. It will be resolved relative to cwd.
+    /// * an absolute path to the configuration file.
+    pub config_file: String,
+    /// Support for Typescript Project References.
+    /// For the type TsconfigReferences
+    ///
+    #[napi(ts_type = "'disabled'| 'auto' | 'Array<string>'")]
+    pub references: Either<String, Vec<String>>,
+}
+
+// #[derive(Debug, Clone)]
+// pub enum TsconfigReferences {
+//     Disabled,
+//     /// Use the `references` field from tsconfig read from `config_file`.
+//     Auto,
+//     /// Manually provided relative or absolute path.
+//     Paths(Vec<String>),
+// }
+
+impl Default for NapiResolveOptions {
+    fn default() -> Self {
+        Self {
+            tsconfig: None,
+            alias: Some(HashMap::new()),
+            alias_fields: Some(vec![]),
+            condition_names: Some(vec![]),
+            description_files: Some(vec!["package.json".into()]),
+            enforce_extension: Some(EnforceExtension::Auto),
+            extension_alias: Some(HashMap::new()),
+            exports_fields: Some(vec![vec!["exports".into()]]),
+            extensions: Some(vec![".js".into(), ".json".into(), ".node".into()]),
+            fallback: Some(HashMap::new()),
+            fully_specified: Some(false),
+            main_fields: Some(vec!["main".into()]),
+            main_files: Some(vec!["index".into()]),
+            modules: Some(vec!["node_modules".into()]),
+            resolve_to_context: Some(false),
+            prefer_relative: Some(false),
+            prefer_absolute: Some(false),
+            restrictions: Some(vec![]),
+            roots: Some(vec![]),
+            symlinks: Some(true),
+            builtin_modules: Some(false),
+        }
+    }
+}
+
+impl Into<oxc_resolver::Restriction> for Restriction {
+    fn into(self) -> oxc_resolver::Restriction {
+        match (self.path, self.regex) {
+            (None, None) => {
+                panic!("Should specify path or regex")
```

Comment:
Can return an Error instead of panic here?

Reply:
populate the `Result` to `ResolverFactory::new`  did make too much sense, the validation should be the caller's duty.

Reply:
You still need to handle the `Result`, I don't think there will be js developer will handle `Result` in the js side.